packages/astro/src/core/errors/errors-data.ts

```diff
@@ -628,6 +628,94 @@ See https://docs.astro.build/en/guides/server-side-rendering/ for more informati
 		code: 3030,
 		message: 'The response has already been sent to the browser and cannot be altered.',
 	},
+
+	/**
+	 * @docs
+	 * @description
+	 * Thrown when a middleware doesn't return anything or doesn't call the `next` function.
+	 *
+	 * For example:
+	 * ```ts
+	 * import {defineMiddleware} from "astro/middleware";
+	 * export const onRequest = defineMiddleware((context, _) => {
+	 * 	// doesn't return anything or doesn't call next
+	 * 	context.locals.someData = false;
+	 * });
+	 * ```
+	 */
+	MiddlewareNoDataOrNextCalled: {
+		title: "The middleware didn't return a response or called `next`",
+		code: 3031,
+		message: 'The middleware needs to return a `Response` object or call the `next` function.',
+	},
```

Comment:
```suggestion
	 * Thrown when the middleware does not return any data or call the `next` function.
	 *
	 * For example:
	 * ```ts
	 * import {defineMiddleware} from "astro/middleware";
	 * export const onRequest = defineMiddleware((context, _) => {
	 * 	// doesn't return anything or call `next`
	 * 	context.locals.someData = false;
	 * });
	 * ```
	 */
	MiddlewareNoDataOrNextCalled: {
		title: "The middleware didn't return a response or call `next`",
		code: 3031,
		message: 'The middleware needs to either return a `Response` object or call the `next` function.',
	},
```
I'm not sure how we want to refer to "a piece of middleware" but "a middleware" doesn't sound quite right.  :smile:

I've suggested "the middleware" but if there's a more descriptive word like "a middleware request" or "a middleware call" to describe the action, or one that describes the thing like "a middleware service", those would be places we could use "a".

This is a little difficult to word because these are two different things that both need to not happen, and they're not easy to describe together using the same words.  :sweat_smile: But I think this is clear for now?